File: src/leetcode/m45/JumpGameII.go
package m45

import "math"

// LeetCode 45
type JumpGameII interface {
	Jump(nums []int) int
}

type Solution1 struct {
}

func (s Solution1) Jump(nums []int) int {
	n := len(nums)
	dp := make([]int, n)
	for i := range dp {
		dp[i] = math.MaxInt
	}
	dp[0] = 0

	for i := 0; i < n; i++ {
		for j := 1; j <= nums[i]; j++ {
			if i+j < n {
				dp[i+j] = minInt(dp[i+j], dp[i]+1)
			}
		}
	}
	return dp[n-1]
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

type Solution2 struct {
}

func (s Solution2) Jump(nums []int) int {
	maxPos := 0
	currEnd := 0
	jumps := 0

	for i := 0; i < len(nums)-1; i++ {
		maxPos = maxInt(maxPos, i+nums[i])

		if i == currEnd {
			jumps++
			currEnd = maxPos
			if currEnd >= len(nums)-1 {
				break
			}
		}
	}
	return jumps
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: src/leetcode/m45/JumpGameII.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.m45;

/**
 * LeetCode 45
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/08/19 09:53
 */
public class JumpGameII {
    public int jump(int[] nums) {
        int jumps = 0;
        int currentEnd = 0;
        int farthest = 0;

        for (int i = 0; i < nums.length - 1; i++) {
            // Update the farthest point that can be reached
            farthest = Math.max(farthest, i + nums[i]);

            // If we've reached the end of the current jump range, we need to jump
            if (i == currentEnd) {
                jumps++;
                currentEnd = farthest;

                // If the farthest point can reach the end, break early
                if (currentEnd >= nums.length - 1) {
                    break;
                }
            }
        }

        return jumps;
    }
}
